source/slang/ir-insts.h

```diff
@@ -610,22 +637,26 @@ struct IRBuilder
         UInt            caseArgCount,
         IRValue* const* caseArgs);
 
-
-    IRDecoration* addDecorationImpl(
-        IRValue*        value,
-        UInt            decorationSize,
-        IRDecorationOp  op);
-
     template<typename T>
     T* addDecoration(IRValue* value, IRDecorationOp op)
     {
-        return (T*) addDecorationImpl(value, sizeof(T), op);
+        assert(getModule());
+        auto decorationSize = sizeof(T);
+        auto decoration = (T*)getModule()->memoryPool.allocZero(decorationSize);
+        new(decoration)T();
```

Comment:
Just as an aside, my general approach is that whenever possible I try to make templated functions as small as possible, and always have them delegate to a non-templated function that does the real work, passing in callbacks, sizes, etc. as needed.

The equivalent in this case would be for us to define something (probably in core) where we can get a plain old function pointer for an operation that constructs/destructs values of type `T`, and then pass that function into `addDecorationImpl` so that we don't have all this code in the templated operations.

Obviously the easier case in the long term *might* be to just drop the destructors and virtual functions so that we don't need to call the constructor any more. That said, there are also some definitely up-sides to having a virtual function table in these objects, when it comes to debugging.